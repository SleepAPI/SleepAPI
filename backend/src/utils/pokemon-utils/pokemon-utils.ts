import { pokemon } from 'sleepapi-common';
import { getBerriesForFilter } from '../berry-utils/berry-utils';

export function getPokemonNames(islands: { cyan: boolean; taupe: boolean; snowdrop: boolean; lapis: boolean }) {
  const allowedBerries = getBerriesForFilter(islands);
  return pokemon.COMPLETE_POKEDEX.filter((pokemon) => allowedBerries.includes(pokemon.berry)).map(
    (pokemon) => pokemon.name
  );
}
