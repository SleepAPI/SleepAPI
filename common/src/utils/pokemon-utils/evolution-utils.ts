import { Pokemon } from '../../domain/pokemon';

export function evolvesFrom(previousForm: Pokemon): Pokemon {
  return {
    ...previousForm,
    remainingEvolutions: previousForm.remainingEvolutions - 1,
  };
}

export function evolvesInto(nextForm: Pokemon): Pokemon {
  return {
    ...nextForm,
    remainingEvolutions: nextForm.remainingEvolutions + 1,
  };
}
