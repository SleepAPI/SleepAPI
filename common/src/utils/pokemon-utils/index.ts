export * from './evolution-utils';
export * from './pokemon-utils';
