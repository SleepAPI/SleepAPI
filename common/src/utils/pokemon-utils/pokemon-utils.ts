import { COMPLETE_POKEDEX, Pokemon } from '../../domain/pokemon';

export function getPokemon(name: string): Pokemon {
  const pkmn = COMPLETE_POKEDEX.find((pokemon) => pokemon.name.toLowerCase() === name.toLowerCase());
  if (!pkmn) {
    throw new Error(`Can't find Pokemon with name ${name}`);
  }
  return pkmn;
}
