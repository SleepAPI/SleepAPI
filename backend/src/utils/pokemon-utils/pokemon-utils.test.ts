import { getPokemonNames } from './pokemon-utils';

describe('getPokemonNames', () => {
  it('shall return all pokemon by default', () => {
    const params = {
      cyan: false,
      taupe: false,
      snowdrop: false,
      lapis: false,
    };
    expect(getPokemonNames(params)).toMatchSnapshot();
  });

  it('shall return all cyan pokemon', () => {
    const params = {
      cyan: true,
      taupe: false,
      snowdrop: false,
      lapis: false,
    };
    expect(getPokemonNames(params)).toMatchInlineSnapshot(`
      [
        "CLEFABLE",
        "DODRIO",
        "FERALIGATR",
        "BLASTOISE",
        "DELIBIRD",
        "WIGGLYTUFF",
        "GOLDUCK",
        "SLOWBRO",
        "VAPOREON",
        "SLOWKING",
        "TOGEKISS",
        "SYLVEON",
      ]
    `);
  });

  it('shall return all taupe pokemon', () => {
    const params = {
      cyan: false,
      taupe: true,
      snowdrop: false,
      lapis: false,
    };
    expect(getPokemonNames(params)).toMatchInlineSnapshot(`
      [
        "ONIX",
        "MAROWAK",
        "TYPHLOSION",
        "GOLEM",
        "DUGTRIO",
        "PUPITAR",
        "CHARIZARD",
        "ARCANINE",
        "FLAREON",
        "SUDOWOODO",
      ]
    `);
  });

  it('shall return all snowdrop pokemon', () => {
    const params = {
      cyan: false,
      taupe: false,
      snowdrop: true,
      lapis: false,
    };
    expect(getPokemonNames(params)).toMatchInlineSnapshot(`
      [
        "RATICATE",
        "HOUNDOOM",
        "VIGOROTH",
        "SLAKING",
        "WALREIN",
        "ABOMASNOW",
        "ABSOL",
        "DITTO",
        "TYRANITAR",
        "KANGASKHAN",
        "PERSIAN",
        "UMBREON",
        "SABLEYE",
        "GLACEON",
      ]
    `);
  });

  it('shall return all lapis pokemon', () => {
    const params = {
      cyan: false,
      taupe: false,
      snowdrop: false,
      lapis: true,
    };
    expect(getPokemonNames(params)).toMatchInlineSnapshot(`
      [
        "PRIMEAPE",
        "MEGANIUM",
        "VENUSAUR",
        "VICTREEBEL",
        "MR_MIME",
        "ESPEON",
        "WOBBUFFET",
        "LUCARIO",
        "LEAFEON",
      ]
    `);
  });

  it('shall return all pokemon for multiple islands', () => {
    const allIslands = {
      cyan: true,
      taupe: true,
      snowdrop: true,
      lapis: true,
    };

    expect(getPokemonNames(allIslands)).toMatchSnapshot();
  });
});
