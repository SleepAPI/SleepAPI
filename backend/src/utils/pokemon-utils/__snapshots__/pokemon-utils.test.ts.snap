// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPokemonNames shall return all pokemon by default 1`] = `
[
  "BUTTERFREE",
  "RATICATE",
  "ARBOK",
  "PIKACHU_HALLOWEEN",
  "RAICHU",
  "CLEFABLE",
  "PRIMEAPE",
  "DODRIO",
  "ONIX",
  "MAROWAK",
  "MEGANIUM",
  "TYPHLOSION",
  "FERALIGATR",
  "STEELIX",
  "HOUNDOOM",
  "VIGOROTH",
  "SLAKING",
  "ALTARIA",
  "BANETTE",
  "WALREIN",
  "ABOMASNOW",
  "VENUSAUR",
  "GOLEM",
  "BLASTOISE",
  "VICTREEBEL",
  "ABSOL",
  "GENGAR",
  "DITTO",
  "DUGTRIO",
  "DELIBIRD",
  "PUPITAR",
  "TYRANITAR",
  "KANGASKHAN",
  "CHARIZARD",
  "PINSIR",
  "TOXICROAK",
  "MR_MIME",
  "PIKACHU_CHRISTMAS",
  "WIGGLYTUFF",
  "PERSIAN",
  "GOLDUCK",
  "ARCANINE",
  "SLOWBRO",
  "VAPOREON",
  "JOLTEON",
  "FLAREON",
  "AMPHAROS",
  "SUDOWOODO",
  "ESPEON",
  "UMBREON",
  "SLOWKING",
  "WOBBUFFET",
  "HERACROSS",
  "SABLEYE",
  "SWALOT",
  "LUCARIO",
  "MAGNEZONE",
  "TOGEKISS",
  "LEAFEON",
  "GLACEON",
  "SYLVEON",
]
`;

exports[`getPokemonNames shall return all pokemon for multiple islands 1`] = `
[
  "RATICATE",
  "CLEFABLE",
  "PRIMEAPE",
  "DODRIO",
  "ONIX",
  "MAROWAK",
  "MEGANIUM",
  "TYPHLOSION",
  "FERALIGATR",
  "HOUNDOOM",
  "VIGOROTH",
  "SLAKING",
  "WALREIN",
  "ABOMASNOW",
  "VENUSAUR",
  "GOLEM",
  "BLASTOISE",
  "VICTREEBEL",
  "ABSOL",
  "DITTO",
  "DUGTRIO",
  "DELIBIRD",
  "PUPITAR",
  "TYRANITAR",
  "KANGASKHAN",
  "CHARIZARD",
  "MR_MIME",
  "WIGGLYTUFF",
  "PERSIAN",
  "GOLDUCK",
  "ARCANINE",
  "SLOWBRO",
  "VAPOREON",
  "FLAREON",
  "SUDOWOODO",
  "ESPEON",
  "UMBREON",
  "SLOWKING",
  "WOBBUFFET",
  "SABLEYE",
  "LUCARIO",
  "TOGEKISS",
  "LEAFEON",
  "GLACEON",
  "SYLVEON",
]
`;
